pepdbagent/modules/project.py

```diff
@@ -970,3 +1076,235 @@ def get_samples(
+                )
+            )
+            if not project_mapping:
+                raise ProjectNotFoundError(
+                    f"No project found for supplied input: '{namespace}/{name}:{tag}'. "
+                    f"Did you supply a valid namespace and project?"
+                )
+
+            sample_dict = self._get_samples_dict(
+                prj_id=project_mapping.id, session=session, with_id=True
+            )
+
+            main_history = session.scalar(
+                select(HistoryProjects)
+                .where(
+                    and_(
+                        HistoryProjects.project_id == project_mapping.id,
+                        HistoryProjects.id == history_id,
+                    )
+                )
+                .order_by(HistoryProjects.update_time.desc())
+            )
+            if not main_history:
+                raise HistoryNotFoundError(
+                    f"No history found for supplied input: '{namespace}/{name}:{tag}'. "
+                    f"Did you supply a valid history id?"
+                )
+
+            changes_mappings = session.scalars(
+                select(HistoryProjects)
+                .where(
+                    and_(
+                        HistoryProjects.project_id == project_mapping.id,
+                    )
+                )
+                .order_by(HistoryProjects.update_time.desc())
+            )
+
+            # Changes mapping is a ordered list from most early to latest changes
+            # We have to loop through each change and apply it to the sample list
+            # It should be done before we found the history_id that user is looking for
+            project_config = None
+
+            for result in changes_mappings:
+                sample_dict = self._apply_history_changes(sample_dict, result)
+
+                if result.id == history_id:
+                    project_config = result.project_yaml
+                    break
+
+        samples_list = order_samples(sample_dict)
+        ordered_samples_list = [sample["sample"] for sample in samples_list]
+
+        if not with_id:
+            for sample in ordered_samples_list:
+                try:
+                    del sample[PEPHUB_SAMPLE_ID_KEY]
+                except KeyError:
+                    pass
+
+        if raw:
+            return {
+                CONFIG_KEY: project_config or project_mapping.config,
+                SAMPLE_RAW_DICT_KEY: ordered_samples_list,
+                SUBSAMPLE_RAW_LIST_KEY: self.get_subsamples(namespace, name, tag),
+            }
+        return peppy.Project.from_dict(
+            pep_dictionary={
+                CONFIG_KEY: project_config or project_mapping.config,
+                SAMPLE_RAW_DICT_KEY: ordered_samples_list,
+                SUBSAMPLE_RAW_LIST_KEY: self.get_subsamples(namespace, name, tag),
+            }
+        )
+
+    @staticmethod
+    def _apply_history_changes(sample_list: dict, change: HistoryProjects) -> dict:
```

Comment:
Shoulld `sample_list: dict` be a `list[dict]` ?

Reply:
I'm not sure, just checking

Reply:
Fixed